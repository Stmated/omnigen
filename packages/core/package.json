{
  "name": "@omnigen/core",
  "version": "0.0.0",
  "private": true,
  "license": "MIT",
  "type": "module",
  "main": "./src/index.ts",
  "types": "./src/index.ts",
  "scripts": {
    "clean": "rm -rf .turbo && rm -rf node_modules && rm -rf dist",
    "dev": "tsc -w",
    "lint": "TIMING=1 eslint src/**/*.ts* --fix",
    "test": "vitest run",
    "test:watch": "vitest"
  },
  "dependencies": {
    "@omnigen/core-log": "workspace:*",
    "@etclabscore/string-json-template-language": "^1.0.0",
    "@open-rpc/schema-utils-js": "^1.16.1",
    "change-case": "^4.1.2",
    "deepmerge": "^4.3.0",
    "js-yaml": "^4.1.0",
    "json-pointer": "^0.6.2",
    "node-fetch": "^2.6.9",
    "object-hash": "^3.0.0",
    "string-similarity": "^4.0.4",
    "ts-json-schema-generator": "^1.2.0",
    "zod": "^3.22.4"
  },
  "devDependencies": {
    "@omnigen/config-typescript": "workspace:*",
    "@omnigen/eslint-config-custom": "workspace:*",
    "@swc/core": "^1.3.32",
    "@types/json-pointer": "^1.0.31",
    "@types/json-schema": "^7.0.15",
    "@types/node": "^18.13.0",
    "@types/node-fetch": "^2.6.2",
    "@types/object-hash": "^2.2.1",
    "@types/string-similarity": "^4.0.0",
    "colors": "^1.4.0",
    "nock": "^13.3.0",
    "typescript": "^5.4.2",
    "type-fest": "^3.6.1"
  }
}
