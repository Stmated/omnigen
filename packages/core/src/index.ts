import {PluginAutoRegistry} from '@omnigen/core-plugin';
import {CommonTransformPlugin, CommonTransform2Plugin, fileWriter, CorePlugin} from './CoreUtilPluginInit.ts';

export * from './ast';
export * from './equality';
export * from './parse';
export * from './util';
export * from './visit';
export * from './write';
export * from './CoreUtilPluginInit.ts';

export default PluginAutoRegistry.register([CorePlugin, CommonTransformPlugin, CommonTransform2Plugin, fileWriter]);
