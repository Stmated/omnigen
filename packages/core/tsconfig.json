{
  "$schema": "https://json.schemastore.org/tsconfig",
  "extends": "@omnigen/config-typescript/build.json",
  "compilerOptions": {
    "outDir": "./dist",
    "rootDir": "./src",
    "baseUrl": "./src"
  }
}
