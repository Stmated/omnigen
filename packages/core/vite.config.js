import {resolve} from 'path';
import {defineConfig} from 'vite';

export default defineConfig({
  build: {
    outDir: 'dist/lib',
    lib: {
      // Could also be a dictionary or array of multiple entry points
      entry: resolve(__dirname, 'src/index.ts'),
      name: '@omnigen/core',
      // the proper extensions will be added
      // fileName: 'core',
    },
    rollupOptions: {
      // make sure to externalize deps that shouldn't be bundled into your library
      external: ['node-fetch', 'promisify', 'node:fs', 'node:url', 'find-up', 'fs/promises', 'fs'],
      output: {
        // Provide global variables to use in the UMD build for externalized deps
        // globals: {
        //   vue: 'Vue',
        // },
      },
      treeshake: 'safest',
    },
  },
  optimizeDeps: {
    exclude: ['sync-fetch'],
  },
});
