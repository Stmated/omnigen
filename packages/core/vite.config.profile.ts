import {resolve} from 'path';
import {type UserConfig} from 'vite';

export default {
  build: {
    outDir: 'dist/profile',
    minify: false,
    terserOptions: {
      compress: false,
      mangle: false,
    },
    lib: {
      // Could also be a dictionary or array of multiple entry points
      entry: resolve(__dirname, 'src/index.profiling.ts'),
      name: '@omnigen/core',
      // the proper extensions will be added
      // fileName: 'core',
    },
    rollupOptions: {
      external: ['node-fetch', 'promisify', 'node:fs', 'node:url', 'find-up'],
      output: {
        // Provide global variables to use in the UMD build for externalized deps
        // globals: {
        //   vue: 'Vue',
        // },
      },
    },
  },
} satisfies UserConfig;
