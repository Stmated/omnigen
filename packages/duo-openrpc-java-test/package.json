{
  "name": "@omnigen/duo-openrpc-java-test",
  "version": "0.0.0",
  "private": true,
  "license": "MIT",
  "type": "module",
  "main": "./src/index.ts",
  "types": "./src/index.ts",
  "scripts": {
    "clean": "rm -rf .turbo && rm -rf node_modules && rm -rf dist",
    "dev": "tsc -w",
    "lint": "TIMING=1 eslint src/**/*.ts* --fix",
    "test": "NODE_OPTIONS=--experimental-vm-modules jest --passWithNoTests",
    "test:watch": "jest --watch --passWithNoTests"
  },
  "jest": {
    "preset": "@omnigen/config-jest"
  },
  "dependencies": {
    "@omnigen/core": "0.0.0",
    "@omnigen/core-log": "0.0.0",
    "@omnigen/core-util": "0.0.0",
    "@omnigen/parser-jsonschema": "0.0.0",
    "@omnigen/parser-openrpc": "0.0.0",
    "@omnigen/target-java": "0.0.0",
    "json-schema": "^0.4.0"
  },
  "devDependencies": {
    "@types/jest": "^29.4.0",
    "@types/json-schema": "^7.0.11",
    "@omnigen/config-jest": "0.0.0",
    "@omnigen/config-typescript": "0.0.0",
    "@omnigen/eslint-config-custom": "0.0.0",
    "@omnigen/utils-test": "0.0.0",
    "@omnigen/utils-test-target-java": "0.0.0",
    "typescript": "beta",
    "jest": "^29.4.1",
    "java-parser": "^2.0.3"
  }
}
