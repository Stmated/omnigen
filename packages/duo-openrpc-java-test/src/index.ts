
export * from './JavaTestUtils.js';
export * from './OpenRpcTestUtils.js';
