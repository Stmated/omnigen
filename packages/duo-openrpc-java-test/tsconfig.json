{
  "$schema": "https://json.schemastore.org/tsconfig",
  "extends": "@omnigen/tsconfig/build.json",
  "compilerOptions": {
    "outDir": "./dist",
    "rootDir": "./src",
    "baseUrl": "./src"
  }
}
