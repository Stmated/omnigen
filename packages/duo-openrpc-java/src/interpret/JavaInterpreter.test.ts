import {DEFAULT_TEST_JAVA_OPTIONS, JavaTestUtils, OpenRpcTestUtils} from '@omnigen/duo-openrpc-java-test';
import {JavaInterpreter} from '@omnigen/target-java';
import {JavaOptions, JavaUtil} from '@omnigen/target-java';
import {OmniTypeKind, OmniUtil} from '@omnigen/core';
import {DEFAULT_OPENRPC_OPTIONS} from '@omnigen/parser-openrpc';
import {OmniPrimitiveBoxMode} from '@omnigen/core';

describe('JavaInterpreter', () => {

  test('ensureBasicParsingDoesNotCrash', async () => {

    const options = DEFAULT_TEST_JAVA_OPTIONS;
    const interpreter = new JavaInterpreter(options);
    const model = await OpenRpcTestUtils.readExample('openrpc', 'petstore-expanded.json', DEFAULT_OPENRPC_OPTIONS, options);
    const interpretation = await interpreter.buildSyntaxTree(model.model, [], model.options);

    expect(interpretation).toBeDefined();

    const compilationUnits = JavaTestUtils.getCompilationUnits(interpretation);
    const classNames = compilationUnits.map(cu => cu.object.name.value).sort();

    expect(classNames).toEqual([
      'CreatePetRequest',
      'CreatePetRequestParams',
      'CreatePetResponse',
      'DeletePetByIdRequest',
      'DeletePetByIdRequestParams',
      'DeletePetByIdResponse',
      'DeletePetByIdResponsePayload',
      'ErrorUnknown',
      'ErrorUnknownError',
      'GetPetByIdRequest',
      'GetPetByIdRequestParams',
      'GetPetByIdResponse',
      'GetPetsRequest',
      'GetPetsRequestParams',
      'GetPetsResponse',
      'JsonRpcError',
      'JsonRpcErrorResponse',
      'JsonRpcRequest',
      'JsonRpcRequestParams',
      'JsonRpcResponse',
      'NewPet',
      'Pet',
    ]);
  });

  test('ensureGenericsAreSpecialized', async () => {

    const options: JavaOptions = {
      ...DEFAULT_TEST_JAVA_OPTIONS,
      generificationBoxMode: OmniPrimitiveBoxMode.WRAP,
    };

    const interpreter = new JavaInterpreter(options);
    const result = await OpenRpcTestUtils.readExample('openrpc', 'primitive-generics.json', DEFAULT_OPENRPC_OPTIONS, options);
    const root = await interpreter.buildSyntaxTree(result.model, [], result.options);

    expect(root).toBeDefined();

    const compilationUnits = JavaTestUtils.getCompilationUnits(root);
    const fileNames = compilationUnits.map(it => it.object.name.value).sort();

    expect(fileNames)
      .toEqual([
        'ErrorUnknown',
        'ErrorUnknownError',
        'GiveIntGetDoubleRequest',
        'GiveIntGetDoubleRequestParams',
        'GiveIntGetDoubleResponse',
        'GiveNumberGetCharRequest',
        'GiveNumberGetCharRequestParams',
        'GiveNumberGetCharResponse',
        'GiveStringGetStringRequest',
        'GiveStringGetStringRequestParams',
        'GiveStringGetStringResponse',
        'JsonRpcError',
        'JsonRpcErrorResponse',
        'JsonRpcRequest',
        'JsonRpcRequestParams',
        'JsonRpcResponse',
        'PrimitiveChar',
        'PrimitiveDouble',
        'PrimitiveInt',
      ]);

    const givIntGetDoubleRequestParams = JavaTestUtils.getCompilationUnit(root, 'GiveIntGetDoubleRequestParams');
    expect(givIntGetDoubleRequestParams.object.extends).toBeDefined();

    const type = givIntGetDoubleRequestParams.object.extends?.type.omniType;
    if (type?.kind != OmniTypeKind.GENERIC_TARGET) throw Error(`Wrong kind`);

    expect(JavaUtil.getClassName(type.source.of)).toEqual('JsonRpcRequestParams');
    expect(type.targetIdentifiers).toHaveLength(1);

    // NOTE: This is currently "REFERENCE" -- but might change later.
    //        If we introduce a new kind of type that is a reference to a custom type created in AST.
    //        This is because it is quite ugly to use "REFERENCE" in case a transformer moved the referenced object.
    expect(type.targetIdentifiers[0].type.kind).toEqual(OmniTypeKind.HARDCODED_REFERENCE);
  });

  test('ensureGenericsAreBoxed', async () => {

    const options: JavaOptions = {
      ...DEFAULT_TEST_JAVA_OPTIONS,
      generificationBoxMode: OmniPrimitiveBoxMode.BOX,
    };

    const interpreter = new JavaInterpreter(options);
    const result = await OpenRpcTestUtils.readExample('openrpc', 'primitive-generics.json', DEFAULT_OPENRPC_OPTIONS, options);
    const root = await interpreter.buildSyntaxTree(result.model, [], result.options);

    expect(root).toBeDefined();

    const compilationUnits = JavaTestUtils.getCompilationUnits(root);
    const fileNames = compilationUnits.map(it => it.object.name.value).sort();

    expect(fileNames)
      .toEqual([
        'ErrorUnknown',
        'ErrorUnknownError',
        'GiveIntGetDoubleRequest',
        'GiveIntGetDoubleRequestParams',
        'GiveIntGetDoubleResponse',
        'GiveNumberGetCharRequest',
        'GiveNumberGetCharRequestParams',
        'GiveNumberGetCharResponse',
        'GiveStringGetStringRequest',
        'GiveStringGetStringRequestParams',
        'GiveStringGetStringResponse',
        'JsonRpcError',
        'JsonRpcErrorResponse',
        'JsonRpcRequest',
        'JsonRpcRequestParams',
        'JsonRpcResponse',
      ]);

    const giveIntGetDoubleRequestParams = JavaTestUtils.getCompilationUnit(root, 'GiveIntGetDoubleRequestParams');
    expect(giveIntGetDoubleRequestParams.object.extends).toBeDefined();

    const type = giveIntGetDoubleRequestParams.object.extends?.type.omniType;
    if (type?.kind != OmniTypeKind.GENERIC_TARGET) throw Error(`Wrong kind: ${OmniUtil.describe(type)}`);

    expect(JavaUtil.getClassName(type.source.of)).toEqual('JsonRpcRequestParams');
    expect(type.targetIdentifiers).toHaveLength(1);
    expect(type.targetIdentifiers[0].type.kind).toEqual(OmniTypeKind.PRIMITIVE);

    // TODO: Check that JsonRpcRequest has the correct generics!!!!!
  });

  test('ensureGenericsAreSkipped', async () => {

    const options: JavaOptions = {
      ...DEFAULT_TEST_JAVA_OPTIONS,
      generificationBoxAllowed: false,
    };

    const interpreter = new JavaInterpreter(options);
    const result = await OpenRpcTestUtils.readExample('openrpc', 'primitive-generics.json', DEFAULT_OPENRPC_OPTIONS, options);
    const root = await interpreter.buildSyntaxTree(result.model, [], result.options);

    expect(root).toBeDefined();

    const compilationUnits = JavaTestUtils.getCompilationUnits(root);
    const classNames = compilationUnits.map(it => it.object.name.value).sort();

    expect(classNames)
      .toEqual([
        'ErrorUnknown',
        'ErrorUnknownError',
        'GiveIntGetDoubleRequest',
        'GiveIntGetDoubleRequestParams',
        'GiveIntGetDoubleResponse',
        'GiveNumberGetCharRequest',
        'GiveNumberGetCharRequestParams',
        'GiveNumberGetCharResponse',
        'GiveStringGetStringRequest',
        'GiveStringGetStringRequestParams',
        'GiveStringGetStringResponse',
        'JsonRpcError',
        'JsonRpcErrorResponse',
        'JsonRpcRequest',
        'JsonRpcRequestParams',
        'JsonRpcResponse',
      ]);

    const giveNumberGetCharResponse = JavaTestUtils.getCompilationUnit(root, 'GiveIntGetDoubleRequestParams');
    expect(giveNumberGetCharResponse.object.extends).toBeDefined();
    expect(giveNumberGetCharResponse.object.extends?.type.omniType?.kind).toEqual(OmniTypeKind.OBJECT);
  });

  test('Interfaces', async () => {

    const result = await OpenRpcTestUtils.readExample('openrpc', 'multiple-inheritance.json');
    const interpreter = new JavaInterpreter(result.options);
    const root = await interpreter.buildSyntaxTree(result.model, [], result.options);

    expect(root).toBeDefined();

    expect(root.children).toHaveLength(23);

    expect(result).toBeDefined();
  });

  test('Mappings', async () => {

    const result = await OpenRpcTestUtils.readExample('openrpc', 'mappings.json');
    const interpreter = new JavaInterpreter(result.options);
    const root = await interpreter.buildSyntaxTree(result.model, [], result.options);

    expect(root).toBeDefined();

    expect(root.children).toHaveLength(17);

    expect(result).toBeDefined();
  });
});
