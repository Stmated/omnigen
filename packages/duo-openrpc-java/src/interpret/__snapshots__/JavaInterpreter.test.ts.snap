// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`JavaInterpreter > ensureBasicParsingDoesNotCrash 1`] = `
[
  "CreatePetRequest",
  "CreatePetRequestParams",
  "CreatePetResponse",
  "DeletePetByIdRequest",
  "DeletePetByIdRequestParams",
  "DeletePetByIdResponse",
  "ErrorUnknown",
  "ErrorUnknownError",
  "GetPetByIdRequest",
  "GetPetByIdRequestParams",
  "GetPetByIdResponse",
  "GetPetsRequest",
  "GetPetsRequestParams",
  "GetPetsResponse",
  "IPetSchema",
  "JsonRpcError",
  "JsonRpcErrorResponse",
  "JsonRpcRequest",
  "JsonRpcRequestParams",
  "JsonRpcResponse",
  "NewPet",
  "Pet",
  "PetDeleted",
  "PetSchema",
]
`;
